### DISCUSSION TITLE
Speech to Text plugin - is a texteditoroperation possible?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#5957] [Q&A] [answered]
Hi all,

@flancast90 and me were working on a speech-to-text TiddlyWiki5 plugin over the last two weeks (https://github.com/flancast90/Speech-To-Text-in-TW5)

We're now [discussing](https://github.com/flancast90/Speech-To-Text-in-TW5/issues/14) how to create a text-editor toolbar button that allows inserting the text recognized by the SpeechRecognition engine at a tiddlers' caret-position

For that, we think about a new texteditoroperation that comes with the plugin

The text-editor button should not trigger the tm-edit-text-operation message
The button should start recording and from the JS - side we would like to dispatch the message when recording ends

The question now is, if we should build a widget that handles the JS side and if it's possible to insert that widget in the buttons' text-field - or what's the best way to inject the wanted functionality in such a custom text-editor button

We would really appreciate if you had some hints for us

best wishes
Simon

**Comment by Jermolene** [comment]
Hi @BurningTreeC the problem here is that text editor messages must do all of their work in a single synchronous function.

If they trigger an asynchronous event handler (as I imagine you would be to retrieve the spoken text), then the context originally passed to the action will no longer be valid. Consider a case such as the edit text widget being deleted in between the button click and the completion of the transcription.

I wonder if this can be approach generically: a simple JS dæmon that waits for a trigger word and then pipes the transcribed text into whatever text input has the focus. It could run as a startup module.

**Reply by BurningTreeC** [reply]
Thanks for replying @Jermolene 
We already have a JS daemon (as a startup module) that waits for trigger words and we could surely add a trigger-word for piping the transcribed text into whatever input has the focus

**Reply by BurningTreeC** [reply]
@Jermolene - what I have difficulty in wrapping my head around is how to pipe the transcribed text into the framed text-editor (I cannot get it using document.activeElement [?])

**Reply by saqimtiaz** [reply]
This might be helpful (taken from [Streams](https://github.com/saqimtiaz/streams/blob/9e5197485a68f95f5dc7ef2473127ee180965d16/plugins/streams/widgets/keyboard-plus.js#L139-L164)):

```javascript
	var activeElement = document.activeElement;
	var selection;
	if(activeElement && activeElement.tagName === "IFRAME") {
		var idoc = activeElement.contentDocument || activeElement.contentWindow.document;
		activeElement = idoc.activeElement;
		selection = idoc.getSelection();
	} else {
		selection = window.getSelection();
	}


	if(window.CodeMirror && document.activeElement.closest(".CodeMirror")) {
		var cm = document.activeElement.closest(".CodeMirror").CodeMirror;
		var cursor = cm.getCursor("start");
		var startRange = cm.getRange({"line":0,"ch":0},{"line":cursor.line,"ch":cursor.ch});
		var selectionStart = startRange.length;
		var selection = cm.getSelection();
		var selectionEnd = selectionStart + selection.length;
		
		variables["selectionStart"] = startRange.length.toString();
		variables["selectionEnd"] = (selectionStart + selection.length).toString();
		variables["selection"] = cm.getSelection().toString();
	} else if(activeElement && selection && ((activeElement.tagName === "INPUT" && activeElement.type === "TEXT") || activeElement.tagName === "TEXTAREA")) {
		variables["selectionStart"] = activeElement.selectionStart.toString();
		variables["selectionEnd"] = activeElement.selectionEnd.toString();
		variables["selection"] = selection.toString();
	}
```

**Reply by BurningTreeC** [reply]
Thank you @saqimtiaz , that's really helpful indeed!

**Reply by saqimtiaz** [reply]
The overall issue of wanting to get access to the widget that created an input field is something I ran into when thinking about [positioning popups at the caret position](https://github.com/Jermolene/TiddlyWiki5/discussions/5585#discussioncomment-760732) as well. It is something I am planning to revisit post 5.2.0

**Reply by saqimtiaz** [reply]
> Thank you @saqimtiaz , that's really helpful indeed!

I recommend you also check that the IFRAME element has the correct class:
`class="tc-edit-texteditor tc-edit-texteditor-body"`

This wasn't necessary because of the scope in which my code runs, but might be for you in case the user has a different iframe focused in the document.

**Comment by BurningTreeC** [comment, marked as answer]
Thank you @Jermolene and @saqimtiaz for you help!
But I'm in a dead end now. When I set the inputs' value and save the tiddler, the changes don't get saved

**Reply by saqimtiaz** [reply]
Share your code please and if possible also a demo wiki.

**Reply by BurningTreeC** [reply]
Here's my code @saqimtiaz 

```
		if(isPiping) {
			var activeElement = document.activeElement;
			var selection;
			if(activeElement && activeElement.tagName === "IFRAME" && activeElement.className === "tc-edit-texteditor tc-edit-texteditor-body") {
				var idoc = activeElement.contentDocument || activeElement.contentWindow.document;
				activeElement = idoc.activeElement;
				selection = idoc.getSelection();
			} else {
				selection = window.getSelection();
			}
			if(window.CodeMirror && document.activeElement.closest(".CodeMirror")) {
				var cm = document.activeElement.closest(".CodeMirror").CodeMirror;
				var cursor = cm.getCursor("start");
				var startRange = cm.getRange({"line":0,"ch":0},{"line":cursor.line,"ch":cursor.ch});
				var selectionStart = startRange.length;
				var selection = cm.getSelection();
				var selectionEnd = selectionStart + selection.length;
				cm.replaceSelection({replacement: fullTranscript});
				isPiping = false;
				fullTranscript = "";
			} else if(activeElement && selection && ((activeElement.tagName.toUpperCase() === "INPUT" && (activeElement.type.toUpperCase() === "TEXT" || activeElement.type.toUpperCase() === "SEARCH")) || activeElement.tagName.toUpperCase() === "TEXTAREA")) {
				var selStart = activeElement.selectionStart,
					selEnd = activeElement.selectionEnd;
				var value = activeElement.value;
				var newText = value.substring(0,selStart) + fullTranscript + value.substring(selEnd,value.length);
				activeElement.value = newText;
				activeElement.setSelectionRange(selStart,selStart + fullTranscript.length);
				isPiping = false;
				fullTranscript = "";
			}
		}
```

**Reply by BurningTreeC** [reply]
A demo wiki:

[tiddlywiki.zip](https://github.com/Jermolene/TiddlyWiki5/files/6992179/tiddlywiki.zip)

- click the microphone button
- focus an input or textarea
- record some text
- say the words "Ok Wiki insert" or "Ok Wiki pipe" to pipe the spoken words into the input or textarea

**Reply by saqimtiaz** [reply]
Trying using `execCommand`, see framed engine for reference (methods relating to executing an operation).

**Reply by saqimtiaz** [reply]
https://github.com/Jermolene/TiddlyWiki5/blob/199ca57f1c9e286c7ff0bf28f9f6a152ca923da5/core/modules/editor/engines/framed.js#L237-L250

**Reply by BurningTreeC** [reply]
Thank you @saqimtiaz , can I do something similar for normal inputs, too?

**Reply by BurningTreeC** [reply]
Sorry, my question was stupid

**Reply by BurningTreeC** [reply]
What I'm really about is how to make those new values of the inputs, textareas and the framed texteditor permanent

**Reply by saqimtiaz** [reply]
Have you tried using execCommand? Do the changes still not persist?

**Reply by BurningTreeC** [reply]
No I haven't tried yet, I'll try. But for inputs like the tiddler title execCommand doesn't work and I would need a way to modify the actually edited tiddler

**Reply by BurningTreeC** [reply]
Update: yes @saqimtiaz , using `execCommand` for the framed textarea works. It makes changes permanent. Thank you for the hint in the right direction.

The missing part is something similar for the normal text inputs

**Reply by BurningTreeC** [reply]
Ok @saqimtiaz , Thank you for your patience!
I've found out that I can `execCommand` for the normal inputs, too! :smile_cat:

**Comment by flancast90** [comment]
@BurningTreeC, just did some of my own tests. ``Document.activeElement`` has some issues because it is an iframe. Instead, we can call something like this monstrosity of code to get the value of the same-domain iframe.
```javascript
document.getElementsByClassName('tc-edit-texteditor tc-edit-texteditor-body')[1].contentWindow.document.body.getElementsByTagName('textarea')[0].innerHTML;
```
We should then be able to access/change/write new content to the iframe textarea. It may not be the "TW Way", but I think it'll work.

**Reply by BurningTreeC** [reply]
Hi @flancast90 , I'm already doing something similar in my tests. The problem is, that we need to update the underlying `editTiddler` which is the tiddler that is being edited by the simple or framed engine (simple.js or framed.js // factory.js)

**Comment by BurningTreeC** [comment]
As of now I'm able to pipe the spoken words into inputs and textareas and the framed texteditor, too.
But if I save such a tiddler, the changes are not saved, which is because the engine (simple.js or framed.js) is not aware of the changes and doesn't update the underlying editTiddler.

Is there a way to update the editTiddler from within our startup script?

**Reply by BurningTreeC** [reply]
The problem is solved, thank you @Jermolene and @saqimtiaz
